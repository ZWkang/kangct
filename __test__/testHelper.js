// import React from 'react';
import { mount, shallow, render } from 'enzyme';

function setUp(type, ...arg) {
  const [Component, props = {}, children = ''] = arg;
  if (~['mount', 'shallow', 'render'].indexOf(type)) {
    return global[type](<Component {...props}>{children}</Component>);
  }
}

export function AutoMath(Component, props = {}, children = '') {
  const wrapper = setUp('render', Component, props, children);
  expect(wrapper).toMatchSnapshot();
}

export function ComponentRenderWithoutError(Component, props = {}, children = '') {
  const wrapper = setUp('mount', Component, props, children);
  expect(() => {
    wrapper.setProps({});
    wrapper.unmount();
  }).not.toThrow();
}

export function childrenShouldBeProps(Component, children) {
  const wrapper = setUp('mount', Component, {}, children);

  expect(wrapper.get(0).props.children).toEqual(children);
}

export function setup(Component, props) {
  const wrapper = setUp(Component, props, null);
  return {
    wrapper,
    props
  };
}
