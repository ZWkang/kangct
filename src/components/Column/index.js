import React from 'react';
import styled, { css } from 'styled-components';
import Flex from '../Flex';
import is from 'styled-is';
const basicSizeWithColumn = 100 / 12;

const columnSizeMap = [
    'is-1',
    'is-2',
    'is-3',
    'is-4',
    'is-5',
    'is-6',
    'is-7',
    'is-8',
    'is-9',
    'is-10',
    'is-11',
    'is-12'
];
const splitMap = ['1-1'];
const defaultMargin = 0;
const defaultColumnsLimit = 12;
const offsetSize = ['left-offset-2'];

const CalculateColumnSize = ({ columnSize = defaultColumnsLimit, point = 1 }) => {
    return Math.round((point / columnSize) * 10 ** 8) / 10 ** 6;
};

const breakPoints = {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
};
const sorts = {
    xs: 0,
    sm: 1,
    md: 2,
    lg: 3,
    xl: 4
};

const GeneratorMedias = (props, filters) => {
    const template = (breakPoint, width = [], offset = []) => {
        if (breakPoint !== 0 && !breakPoint) return;
        return `
      @media (min-width: ${breakPoint}px) {
        ${width.join('')}
        ${offset.join('')}
      }
    `;
    };
    const mediaqueryString = filters
        .sort((prev, next) => sorts[prev.slice(0, 2)] - sorts[next.slice(0, 2)])
        .map((item) => {
            // md-is-1-3
            if (/\-is\-\d+\-\d+$/.test(item)) {
                let [type, ...rest] = item.split('-');
                const size = rest.join('-');
                const width = CalculateDefaultFlexColumn({ [size]: true }, [size]);
                return template(breakPoints[type], width);
            }
            // md-is-3
            if (/\-is\-\d+$/.test(item)) {
                let [type, ...rest] = item.split('-');
                const size = rest.join('-');
                const width = CalculateDefaultColumnCall({ [size]: true });
                return template(breakPoints[type], width);
            }
            // md-left-offset-3
            if (/\-.*\-offset\-\d+$/.test(item)) {
              let [type, ...rest] = item.split('-');
              const size = rest.join('-');
              const offset = CalculateOffsetSize({ [size]: true }, [size]);
              return template(breakPoints[type], '', offset);
            } 
            // md-left-offset-2-3
            if (/\-.*\-offset\-\d+\-\d+$/.test(item)) {
                let [type, ...rest] = item.split('-');
                const size = rest.join('-');
                const width = CalculateDefaultFlexColumn({ [size]: true });
                return template(breakPoints[type], width);
            }

        })
        .filter(Boolean)
        .join('');
    return css`
        ${mediaqueryString}
    `;
};

//   return `flex: none; width: ${Number(trueValue) * basicSizeWithColumn}%;`
// }
const CalculateDefaultColumnCall = (props) => {
    const result = columnSizeMap.filter((v) => {
        return Boolean(props[v]) === true;
    });
    if (!result || result.length === 0) return '';
    let size = result[result.length - 1];
    let [, point] = size.split('-');
    let coulumnProsent = CalculateColumnSize({ columnSize: defaultColumnsLimit, point: point });
    return css`
        flex: none;
        width: ${coulumnProsent}%;
    `;
};
const RegexpFilter = (val) => (cb) => (props) =>
    cb(props, Object.keys(props).filter((item) => val.test(item)));

const CalculateOffsetSize = (props, filters) => {
    if (filters.length <= 0) return;

    const readySplit = filters[filters.length - 1];

    const [direction, , point, columnSize] = readySplit.split('-');

    let blockDirection = 'margin-right';
    console.log(direction);
    if (direction === 'left') {
        blockDirection = 'margin-left';
    }

    let coulumnProsent = CalculateColumnSize({ columnSize: columnSize, point: point });
    return `
    ${blockDirection} :${coulumnProsent}%;
  `;
};

const CalculateDefaultFlexColumn = (props, val) => {
    if (val.length === 0) return;
    const readySplit = val[val.length - 1];
    const [, point, columnSize] = readySplit.split('-');

    let coulumnProsent = CalculateColumnSize({ columnSize: columnSize, point: point });
    return css`
        flex: none;
        width: ${coulumnProsent}%;
    `;
};

const getProps = (val) => (props) => props[val];
const Column = styled.div`
  position: relative;
  box-sizing: border-box;
  display: inline-block;
  /* flex-basis: 0;
  flex-grow: 1;
  flex-shrink: 1;
  padding: .75rem; */
  min-height: 1px;
  padding-top: .75rem;
  padding-bottom: .75rem;
  float: left;
  
  &:after{
    clear: both;
    content: " ";
    display: table;
  }
  
  ${CalculateDefaultColumnCall}
  ${is('gutter')`
    padding-left: ${getProps('gutter')}
    padding-right: ${getProps('gutter')}
  `}
  ${is('border')`
    border: 1px solid #ddd;
  `}
  ${RegexpFilter(/^is-\d{0,2}\-\d{0,2}/i)(CalculateDefaultFlexColumn)}
  ${RegexpFilter(/^(left|right)-offset-\d{1,5}(-\d{1,5})?$/i)(CalculateOffsetSize)}
  ${RegexpFilter(/(xs|sm|md|lg|xl)/i)(GeneratorMedias)}
  
`;

Column.Row = styled.div`
  display: block;
`
/**
 * @component
 */
export default Column;