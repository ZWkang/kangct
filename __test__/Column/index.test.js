import Column from '@app/Column/index';
import { ComponentRenderWithoutError, AutoMath } from '../testHelper.js';

import 'jest-styled-components';

const TestCase = () => (
  <Column.Row>
    <Column md-is-6>md-is-6</Column>
    <Column md-is-6>md-is-6</Column>
  </Column.Row>
);

describe('Column', () => {
  describe('Column rendering', () => {
    it('without error', () => {
      ComponentRenderWithoutError(TestCase, {});
    });
    it('autoMatch snapshot', () => {
      AutoMath(TestCase, {}, null);
    });
  });
  describe('Column Interactions', () => {
    it('test column column size', () => {
      const wrapper = mount(
        <Column.Row>
          <Column is-6>is-6</Column>
          <Column is-6>is-6</Column>
          <Column is-1>is-1</Column>
          <Column is-2>is-2</Column>
          <Column is-3>is-3</Column>
          <Column is-4>is-4</Column>
          <Column is-5>is-5</Column>
          <Column is-6>is-6</Column>
          <Column is-7>is-7</Column>
          <Column is-8>is-8</Column>
          <Column is-9>is-9</Column>
          <Column is-10>is-10</Column>
          <Column is-11>is-11</Column>
          <Column is-12>is-12</Column>
        </Column.Row>
      );

      // expect(wrapper).
      // console.log(wrapper.debug());
      expect(wrapper.find(Column)).toHaveLength(14);
      expect(
        wrapper
          .find(Column)
          .at(0)
          .text()
      ).toEqual('is-6');
      expect(wrapper.find(Column)).toHaveStyleRule('flex', 'none');
      expect(wrapper.find(Column)).toHaveStyleRule('width', '50%');

      expect(wrapper.find(Column).at(9)).toHaveStyleRule('width', '66.666667%');
      expect(
        wrapper
          .find(Column)
          .at(9)
          .text()
      ).toEqual('is-8');
    });

    it('test column offset size', () => {
      const wrapper = mount(
        <Column.Row>
          <Column is-6>is-6</Column>
          <Column left-offset-3 is-3>
            offset-3
          </Column>
        </Column.Row>
      );

      expect(
        wrapper
          .find(Column)
          .at(1)
          .text()
      ).toEqual('offset-3');

      expect(wrapper.find(Column).at(1)).toHaveStyleRule('margin-left', '25%');
    });
    it('test Column gutter', () => {
      const wrapper = mount(<Column gutter={'8px'}></Column>);
      expect(wrapper).toHaveStyleRule('padding-left', '8px');
      expect(wrapper).toHaveStyleRule('padding-right', '8px');

      wrapper.setProps({
        gutter: '10px'
      });
      expect(wrapper).toHaveStyleRule('padding-left', '10px');
      expect(wrapper).toHaveStyleRule('padding-right', '10px');
    });
    it('test Column border', () => {});
  });

  describe('Column Interactions', () => {});
});
