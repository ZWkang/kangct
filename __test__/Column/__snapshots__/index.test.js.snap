// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Column rendering autoMatch snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH SLoFM",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa eBdSjf",
        },
        "children": Array [
          Object {
            "data": "md-is-6",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "sc-bdVaJa eBdSjf",
          },
          "children": Array [
            Object {
              "data": "md-is-6",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa eBdSjf",
        },
        "children": Array [
          Object {
            "data": "md-is-6",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "sc-bdVaJa eBdSjf",
          },
          "children": Array [
            Object {
              "data": "md-is-6",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
